### PR TITLE
Add API to add/remove nodes, as well as get the node-name for a key

### PR DESCRIPTION
**Pull request by Swatinem**
Also, add a proper test that actually shows the invariants of consistent hashing.

This is a better alternative to #53

### CODE REVIEW

File: rust/src/ketama.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    ranking: Vec<NodeRank>,
+
+    /// A reusable scratch buffer to format node hash keys into.
+    hash_buf: String,
+}
+
+/// A Nodes `rank` in the main [`KetamaPool::ranking`] list.
+struct NodeRank {
+    value: u32,
+    index: u32,
```
Comment (loewenheim) [inline rust/src/ketama.rs]: These could use docstrings

File: rust/src/ketama.rs [hunk 2]
```diff
@@ -1,27 +1,64 @@
...
     /// The list of Servers, sorted by their ranking value.
-    ranking: Vec<ServerRank>,
+    ranking: Vec<NodeRank>,
+
+    /// A reusable scratch buffer to format node hash keys into.
+    hash_buf: String,
 }
 
-struct ServerRank {
+/// A Nodes `rank` in the main [`KetamaPool::ranking`] list.
```
Comment (loewenheim) [inline rust/src/ketama.rs:16]:
```suggestion
/// A node's `rank` in the main [`KetamaPool::ranking`] list.
```

File: rust/src/ketama.rs [hunk 3]
```diff
@@ -40,38 +77,82 @@ impl KetamaPool {
         };
         self.ranking[ranking_idx % self.ranking.len()].index as usize
     }
-}
 
-const POINTS_PER_HASH: usize = 4;
-const POINTS_PER_SERVER: usize = 40;
+    fn update_node_ranking(&mut self) {
+        self.ranking.clear();
```
Comment (loewenheim) [inline rust/src/ketama.rs:82]: Doesn't this mean you have to recreate the entire ranking on every addition and deletion? Maybe I'm misunderstanding something, but I thought the point of consistent hashing was not to do this.

**Review by loewenheim — APPROVED** [review]

File: rust/src/ketama.rs (hunk 3 above)
Comment (Swatinem) [reply rust/src/ketama.rs]: I added the comment noting that we can incrementally `add` if we really wanted to, but its probably not worth it.

File: bindings/src/ketama.rs [hunk 4]
```diff
@@ -19,6 +19,6 @@ impl KetamaPool {
     }
 
     fn get_slot(&self, key: &str) -> usize {
-        self.0.get_slot(key)
+        self.0.get_node_idx(key)
```
Comment (markstory) [inline bindings/src/ketama.rs:22]: Wouldn't we also need a way to get the server name back to python? We'll need the server name to satisfy the `get_node` interface for the hasher.

File: rust/src/ketama.rs [hunk 5]
```diff
@@ -1,27 +1,66 @@
...
+        self.update_node_ranking();
+    }
+
+    /// Returns the node name which will host the given `key`.
     ///
-    /// The "slot" here is an index into the origin list of keys this pool was constructed with.
-    pub fn get_slot(&self, key: &str) -> usize {
-        if self.ranking.len() == 1 {
+    /// Panics if no node has been added to this pool.
+    pub fn get_node(&self, key: &str) -> &str {
```
Comment (markstory) [inline rust/src/ketama.rs:56]: I think we'll need this in the python binding as well.

**Review by anonrig — COMMENTED** [review]
Why is it a "better alternative"? Can you elaborate?

**Comment by Swatinem** [comment]
> Why is it a "better alternative"? Can you elaborate?

Well:
- It does not have a completely broken `add` and `remove` method for one.
- It does not do 160 completely unnecessary string copies for each node.
- It does not inflate the hash ring by `24 * 160` bytes per node.
- It actually has a *useful* test that clearly demonstrates the desired effects of consistent hashing.
- ☝🏻 a test BTW which also proves that add / remove are not completely broken.

It is missing python bindings for the added methods however ;-)
Although I’m pretty sure we don’t need those though, as the list of nodes is static and does not change at runtime.

**Pull request merged**